File: src/main/java/me/knighthat/innertube/request/Request.java
package me.knighthat.innertube.request;

import lombok.Builder;
import lombok.Value;
import me.knighthat.innertube.request.body.RequestBody;
import org.intellij.lang.annotations.MagicConstant;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Unmodifiable;

import java.util.List;
import java.util.Map;

// Ripped from NewPipeExtractor
@Value
@Builder(builderClassName = "Builder")
public class Request {

    // START: Static fields/functions
    @NotNull
    public static final String GET  = "GET";
    @NotNull
    public static final String POST = "POST";
    // END: Static fields/functions

    @NotNull
    @MagicConstant(valuesFromClass = Request.class)
    String                    httpMethod;
    @NotNull
    @Unmodifiable
    Map<String, List<String>> headers;
    @NotNull  String      url;
    @Nullable RequestBody dataToSend;   // Should only be null when [httpMethod] is "GET"
}

File: src/main/kotlin/me/knighthat/innertube/request/Request.kt
package me.knighthat.innertube.request

import me.knighthat.innertube.request.Request.Companion.GET
import me.knighthat.innertube.request.Request.Companion.POST
import me.knighthat.innertube.request.body.RequestBody
import org.intellij.lang.annotations.MagicConstant

/**
 * Simple class represents the data to be sent.
 *
 * Params' description is copied from [mdn web docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/)
 *
 * [dataToSend] should be in Json format.
 *
 * @param httpMethod Indicate the purpose of the request, use [GET] or [POST]
 * @param headers Pass additional information
 * @param url Where to send this request to
 * @param dataToSend Content of request, can be `null` if [httpMethod] is **GET**
 */
// Ripped from NewPipeExtractor
data class Request(
    @MagicConstant(valuesFromClass = Request::class)
    val httpMethod: String? = null,
    val headers: Map<String, List<String>>,
    val url: String,
    val dataToSend: RequestBody?
) {

    companion object {

        const val GET: String = "GET"
        const val POST: String = "POST"
    }
}
